src/egraph.rs

```diff
@@ -510,19 +547,35 @@ impl<L: Language, N: Analysis<L>> EGraph<L, N> {
         true
     }
 
+    fn perform_to_union(&mut self) {
+        while !self.to_union.is_empty() {
+            let mut current = vec![];
+            std::mem::swap(&mut self.to_union, &mut current);
+            for (id1, id2, rule) in current.into_iter() {
+                self.perform_union(id1, id2, rule.map(|r| Justification::Rule(r)));
+            }
+        }
+    }
+
     #[inline(never)]
     fn process_unions(&mut self) -> usize {
+        self.perform_to_union();
+
         let mut n_unions = 0;
 
         while !self.pending.is_empty() {
-            while let Some((mut node, class)) = self.pending.pop() {
-                node.update_children(|id| self.find_mut(id));
-                if let Some(memo_class) = self.memo.insert(node, class) {
-                    let (_, did_something) = self.union(memo_class, class);
-                    n_unions += did_something as usize;
+            while !self.pending.is_empty() {
```

Comment:
Here is the big change to rebuilding- we have to restore congruence completely before doing analysis stuff